Add retry for docker pull to deal with occasional errors from ghcr

peterbroadhurst opened the pull request:
Split `docker-compose pull` into a separate step with retries, as a simple way to address https://github.com/hyperledger/firefly/issues/244

nguyer commented:
Closing and re-opening to trigger the DCO check again

peterbroadhurst commented:
Tried draft/un-draft too. Think we might need a comment based confirmation of DCO in this case.

nguyer commented on internal/stacks/stack_manager.go:
```diff
@@ -360,6 +360,11 @@ func (s *StackManager) runStartupSequence(workingDir string, verbose bool, first
+	if err := docker.RunDockerComposeCommandWithRetry(workingDir, verbose, verbose, "pull"); err != nil {
+		return err
+	}
+
```
We already do a pull (if `--no-pull` is not set) here:

https://github.com/hyperledger/firefly-cli/blob/e645f531923524119451d1b94d222fa87a497f24/internal/stacks/stack_manager.go#L487-L492

Should we just change that existing call to `RunDockerComposeCommandWithRetry`?

peterbroadhurst replied:
Aha, this is a great point… but what it points at (which I totally missed) is the problem is the e2e test is doing a no-pull 🤦‍♂️

So 👍 to moving the retry on the existing pull, but it means this won’t actually fix the issue.

We might need to punt this to after your manifest work, and have a new behavior for the e2e test where it no long specifies no-pull, but instead uses a different (local) image name.

peterbroadhurst replied:
I've made the change on this PR, and am up for working on the follow-on change to generate a special manifest in the e2e test code once the manifest work is ready

peterbroadhurst commented:
We need to extend this to use `docker-compose pull --filter "label=xyz"` to avoid local images being pulled

peterbroadhurst commented:
My earlier assertion that the `repo` part of the docker image would be enough to stop docker-compose pulling it, seems to be proven false:
https://github.com/docker/compose/issues/3660

This leaves us in a bit of a quandary (as many others on that thread are).
We can:

- Use `--ignore-pull-failures` and still do a pull to hopefully work around occasional ghcr failures
- Move the retry to the whole `docker-compose start`

The latter worries me, because it could cause us to miss actual startup error conditions like https://github.com/hyperledger/firefly/issues/254

peterbroadhurst commented:
Ok - think this latest approach does what we need

- Added a `ff pull <STACK> --retires [X]` command for e2e to use
- Added a `local: true` option to the manifest, to allow a particular image to be skipped pulling
- Automatically add `local: true` for the case the CLI auto-adds FireFly core today

Then a small tweak to the E2E is to do a `pull -retires 3` before a `start`

nguyer reviewed (approved):
I tested this version of the CLI with the corresponding core PR https://github.com/hyperledger/firefly/pull/251 and all E2E tests passed locally! :smiley:
